### PR TITLE
source-postgres : Add logic to handle xmin wraparound

### LINKED ISSUE #25753: Implement wraparound sync logic for Xmin

### PR DESCRIPTION
**Pull request by akashkulk**
Closes https://github.com/airbytehq/airbyte/issues/25753

Wraparound logic is roughly : 
1. If the `xmin` value has wrapped around twice, we'll need to perform a full sync
2. If the `xmin` value has wrapped around once, we'll need to issue a `SELECT * FROM table WHERE xmin::text::bigint >= prev_value OR xmin::text::bigint < current_val` quert to ensure we get all of the changes
See https://docs.google.com/document/d/1pp3Yr0m_y575hIqhvVNUDT7RjldXfGJbR-wMifOA38g/edit#heading=h.ce7nf543uo37 for more details

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  source-postgres-strict-encrypt test report (commit [`361eac83fc`](https://github.com/airbytehq/airbyte/pull/27384/commits/361eac83fce706dd71e26aadbde3b78a1c61fa2f)) - ❌

⏲️  Total pipeline duration: 903 seconds

| Step                                                                                  | Result   |
|:--------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                       | ✅        |
| Connector version increment check.                                                    | ❌        |
| QA checks                                                                             | ✅        |
| Build connector tar                                                                   | ✅        |
| Build source-postgres-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                            | ✅        |
| Integration tests                                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274371753)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`361eac83fc`](https://github.com/airbytehq/airbyte/pull/27384/commits/361eac83fce706dd71e26aadbde3b78a1c61fa2f)) - ❌

⏲️  Total pipeline duration: 6 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Build connector tar                                                   | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274371753)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`361eac83fc`](https://github.com/airbytehq/airbyte/pull/27384/commits/361eac83fce706dd71e26aadbde3b78a1c61fa2f)) - ❌

⏲️  Total pipeline duration: 6 seconds

| Step                                                                                 | Result   |
|:-------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                      | ✅        |
| Connector version increment check.                                                   | ❌        |
| QA checks                                                                            | ✅        |
| Build connector tar                                                                  | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274371753)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`361eac83fc`](https://github.com/airbytehq/airbyte/pull/27384/commits/361eac83fce706dd71e26aadbde3b78a1c61fa2f)) - ❌

⏲️  Total pipeline duration: 9 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274371753)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`5b99348864`](https://github.com/airbytehq/airbyte/pull/27384/commits/5b99348864c8424f5bba11fe6dd968f6d2ad0ff6)) - ❌

⏲️  Total pipeline duration: 1721 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ✅        |
| Build source-postgres docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                             | ✅        |
| Integration tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274604663)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`5b99348864`](https://github.com/airbytehq/airbyte/pull/27384/commits/5b99348864c8424f5bba11fe6dd968f6d2ad0ff6)) - ❌

⏲️  Total pipeline duration: 457 seconds

| Step                                                                                 | Result   |
|:-------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                      | ✅        |
| Connector version increment check.                                                   | ❌        |
| QA checks                                                                            | ✅        |
| Build connector tar                                                                  | ✅        |
| Build source-alloydb-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                           | ✅        |
| Integration tests                                                                    | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274604663)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres-strict-encrypt test report (commit [`5b99348864`](https://github.com/airbytehq/airbyte/pull/27384/commits/5b99348864c8424f5bba11fe6dd968f6d2ad0ff6)) - ❌

⏲️  Total pipeline duration: 732 seconds

| Step                                                                                  | Result   |
|:--------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                       | ✅        |
| Connector version increment check.                                                    | ❌        |
| QA checks                                                                             | ✅        |
| Build connector tar                                                                   | ✅        |
| Build source-postgres-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                            | ✅        |
| Integration tests                                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274604663)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`5b99348864`](https://github.com/airbytehq/airbyte/pull/27384/commits/5b99348864c8424f5bba11fe6dd968f6d2ad0ff6)) - ❌

⏲️  Total pipeline duration: 460 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Build connector tar                                                   | ✅        |
| Build source-alloydb docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                            | ✅        |
| Integration tests                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5274604663)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java [hunk 1]
```diff
@@ -138,24 +138,75 @@ private PreparedStatement createXminQueryStatement(
...
+    if (shouldPerformFullSync(prevRunXminStatus, currentXminStatus)) {
+      final String sql = String.format("SELECT %s FROM %s", wrappedColumnNames, fullTableName);
+      return connection.prepareStatement(sql);
+    } else if (isSingleWraparound(prevRunXminStatus, currentXminStatus)) {
       // The xmin state that we save represents the lowest XID that is still in progress. To make sure we
-      // don't miss
-      // data associated with the current transaction, we have to issue an >=
+      // don't miss data associated with the current transaction, we have to issue an >=. Because of the wraparound, the changes prior to the
+      // end xmin xid value must also be captured.
+      final String sql = String.format("SELECT %s FROM %s WHERE xmin::text::bigint >= ? OR xmin::text::bigint < ?",
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java:166]: What is the expectation in terms of large table?
Are two logical > < going to take twice as much time?

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java [hunk 2]
```diff
@@ -138,24 +138,75 @@ private PreparedStatement createXminQueryStatement(
...
+      return currentXminStatus.getNumWraparound() - 0 == 1;
+    }
+  }
+
+  // Detects whether source Postgres DB has undergone multiple wraparound events between syncs.
+  static boolean shouldPerformFullSync(final XminStatus prevRunXminStatus, final XminStatus currentXminStatus) {
+    if (prevRunXminStatus != null) {
+      return currentXminStatus.getNumWraparound() - prevRunXminStatus.getNumWraparound() >= 2;
+    } else {
+      return currentXminStatus.getNumWraparound() - 0 >= 2;
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java:209]: `- 0`?

**Review by subodh1810 — CHANGES_REQUESTED** [review]
Can we make this verification In postgres source itself? Before we initialise the iterator, go through the saved states and verify that the xmin value is valid. If we do it that way then we can decide which iterator to use (ctid or xmin). For this PR we can just pass an empty state if the xmin is no longer valid.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java (hunk 1 above)
Comment (akashkulk) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java:166]: No, it still requires just a full table scan. The query optimizer ends up checking for a single filter condition.

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java (hunk 2 above)
Comment (akashkulk) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java]: Refactored this to check for `null` previous state upfront. Please take a look

**Comment by akashkulk** [comment]
> 
Isn't that handled in your [PR](https://github.com/airbytehq/airbyte/pull/27302/files#diff-f14bc8e9216a505150004f54d41ca7aeeba9a9460fc4a660b11225dd7abb274f)? If there is no previous `xmin` state, then we automatically start the CTID sync. 

However, the other case (which you might already be referring to) is the case where we have a double wraparound. You're right, in that case, we do want to switch to a CTID sync. However, in practice I think this will be _extremely_ rare and would indicate a larger problem with the user's DB, so can punt to if/when we see a lot of these cases

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`60626d801f`](https://github.com/airbytehq/airbyte/pull/27384/commits/60626d801f6ae71bdc994003be402eaacd31a57b)) - ❌

⏲️  Total pipeline duration: 1713 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ✅        |
| Build source-postgres docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                             | ✅        |
| Integration tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5282951984)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres-strict-encrypt test report (commit [`60626d801f`](https://github.com/airbytehq/airbyte/pull/27384/commits/60626d801f6ae71bdc994003be402eaacd31a57b)) - ❌

⏲️  Total pipeline duration: 729 seconds

| Step                                                                                  | Result   |
|:--------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                       | ✅        |
| Connector version increment check.                                                    | ❌        |
| QA checks                                                                             | ✅        |
| Build connector tar                                                                   | ✅        |
| Build source-postgres-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                            | ✅        |
| Integration tests                                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5282951984)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`7e7eb9539e`](https://github.com/airbytehq/airbyte/pull/27384/commits/7e7eb9539e60ea2152273b81edd6925105929a0d)) - ❌

⏲️  Total pipeline duration: 279 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Build connector tar                                                   | ✅        |
| Build source-alloydb docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                            | ✅        |
| Integration tests                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5283002600)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`7e7eb9539e`](https://github.com/airbytehq/airbyte/pull/27384/commits/7e7eb9539e60ea2152273b81edd6925105929a0d)) - ❌

⏲️  Total pipeline duration: 1093 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ✅        |
| Build source-postgres docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                             | ✅        |
| Integration tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5283002600)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`7e7eb9539e`](https://github.com/airbytehq/airbyte/pull/27384/commits/7e7eb9539e60ea2152273b81edd6925105929a0d)) - ❌

⏲️  Total pipeline duration: 288 seconds

| Step                                                                                 | Result   |
|:-------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                      | ✅        |
| Connector version increment check.                                                   | ❌        |
| QA checks                                                                            | ✅        |
| Build connector tar                                                                  | ✅        |
| Build source-alloydb-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                           | ✅        |
| Integration tests                                                                    | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5283002600)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`51373e0d9a`](https://github.com/airbytehq/airbyte/pull/27384/commits/51373e0d9ab2a16526934642c4a57939ccc10a63)) - ❌

⏲️  Total pipeline duration: 335 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Build connector tar                                                   | ✅        |
| Build source-alloydb docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                            | ✅        |
| Integration tests                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5292012646)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`51373e0d9a`](https://github.com/airbytehq/airbyte/pull/27384/commits/51373e0d9ab2a16526934642c4a57939ccc10a63)) - ❌

⏲️  Total pipeline duration: 1255 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ✅        |
| Build source-postgres docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                             | ✅        |
| Integration tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5292012646)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres-strict-encrypt test report (commit [`51373e0d9a`](https://github.com/airbytehq/airbyte/pull/27384/commits/51373e0d9ab2a16526934642c4a57939ccc10a63)) - ❌

⏲️  Total pipeline duration: 533 seconds

| Step                                                                                  | Result   |
|:--------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                       | ✅        |
| Connector version increment check.                                                    | ❌        |
| QA checks                                                                             | ✅        |
| Build connector tar                                                                   | ✅        |
| Build source-postgres-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                            | ✅        |
| Integration tests                                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5292012646)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`51373e0d9a`](https://github.com/airbytehq/airbyte/pull/27384/commits/51373e0d9ab2a16526934642c4a57939ccc10a63)) - ❌

⏲️  Total pipeline duration: 327 seconds

| Step                                                                                 | Result   |
|:-------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                      | ✅        |
| Connector version increment check.                                                   | ❌        |
| QA checks                                                                            | ✅        |
| Build connector tar                                                                  | ✅        |
| Build source-alloydb-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                           | ✅        |
| Integration tests                                                                    | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5292012646)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres test report (commit [`387bd5c266`](https://github.com/airbytehq/airbyte/pull/27384/commits/387bd5c266eac3f48792b042718897142ede65c1)) - ❌

⏲️  Total pipeline duration: 1287 seconds

| Step                                                                   | Result   |
|:-----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres/metadata.yaml | ✅        |
| Connector version semver check.                                        | ✅        |
| Connector version increment check.                                     | ❌        |
| QA checks                                                              | ✅        |
| Build connector tar                                                    | ✅        |
| Build source-postgres docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                             | ✅        |
| Integration tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5294134197)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres test
```

**Comment by octavia-squidington-iii** [comment]
##  source-postgres-strict-encrypt test report (commit [`387bd5c266`](https://github.com/airbytehq/airbyte/pull/27384/commits/387bd5c266eac3f48792b042718897142ede65c1)) - ❌

⏲️  Total pipeline duration: 506 seconds

| Step                                                                                  | Result   |
|:--------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-postgres-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                       | ✅        |
| Connector version increment check.                                                    | ❌        |
| QA checks                                                                             | ✅        |
| Build connector tar                                                                   | ✅        |
| Build source-postgres-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                            | ✅        |
| Integration tests                                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5294134197)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-postgres-strict-encrypt test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb test report (commit [`387bd5c266`](https://github.com/airbytehq/airbyte/pull/27384/commits/387bd5c266eac3f48792b042718897142ede65c1)) - ❌

⏲️  Total pipeline duration: 325 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Build connector tar                                                   | ✅        |
| Build source-alloydb docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                            | ✅        |
| Integration tests                                                     | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5294134197)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb test
```

**Comment by octavia-squidington-iii** [comment]
##  source-alloydb-strict-encrypt test report (commit [`387bd5c266`](https://github.com/airbytehq/airbyte/pull/27384/commits/387bd5c266eac3f48792b042718897142ede65c1)) - ❌

⏲️  Total pipeline duration: 329 seconds

| Step                                                                                 | Result   |
|:-------------------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-alloydb-strict-encrypt/metadata.yaml | ✅        |
| Connector version semver check.                                                      | ✅        |
| Connector version increment check.                                                   | ❌        |
| QA checks                                                                            | ✅        |
| Build connector tar                                                                  | ✅        |
| Build source-alloydb-strict-encrypt docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                           | ✅        |
| Integration tests                                                                    | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5294134197)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-alloydb-strict-encrypt test
```

**Review by subodh1810 — CHANGES_REQUESTED** [review]
What's bothering me is the fact that we are deciding to do an initial sync inside class `PostgresXminHandler`. We should make that call before so that the logic can be extended to ctid sync. Like in my PR https://github.com/airbytehq/airbyte/pull/27302 when we categorise the syncs that need to go through xmin or ctid sync, we should add this logic that if a stream has xmin state but the xmin value has gone through wrap around, add that stream to the list of streams that need to go through ctid sync. I would have been fine to go ahead and merge this PR if the work related to ctid would have been a week or so behind but the other PR is also ready so I think we should build this logic on top of the that PR. What do you think?

**Comment by subodh1810** [comment]
@akashkulk here is a PR https://github.com/airbytehq/airbyte/pull/27466/files which is built of top of ctid PR and triggers full sync in case of a double wrap around via ctid. Please take a look

**Comment by rodireich** [comment]
Should this be closed for the PR Subodh created? #27466

**Comment by akashkulk** [comment]
> Should this be closed in favor for the PR Subodh created? #27466

Yes, closing for now

**Pull request closed without merging**